src/funcx_common/redis_task.py

```diff
@@ -0,0 +1,188 @@
+
+    Loading:
+      Read a task from storage using the `load()` classmethod, e.g.
+      >>> RedisTask.load(redis_client, "foo_id")
+
+    Creation and Loading will each check `RedisTask.exists` when invoked. If the task_id
+    already exists on creation, or does not exist on load, a ValueError will be raised.
+    For example:
+      >>> RedisTask.load(redis_client, "foo_id")
+      Traceback (most recent call last):
+        ...
+      ValueError: Cannot load task foo_id: does not exist
+      >>> RedisTask(redis_client, "foo_id")  # ok
+      >>> RedisTask(redis_client, "foo_id")
+      Traceback (most recent call last):
+        ...
+      ValueError: Conflict. Cannot create task foo_id: already exists
+
+    This class provides various task fields as descriptors via RedisField, and is
+    responsible for de/serializing various data from/to hstore in Redis.
+
+    There are several elements of this pattern of use which need to be fixed. It is
+    important to be aware of the following:
+    - there is currently no use of Redis transactions, so nothing is atomic
+    - Each time a field descriptor is accessed, it is read, returned, and discarded.
+      Reading a field multiple times, even in a single python statement, is vulnerable
+      to data races
+    - no field requirements or validity are enforced -- reading a field can raise an
+      error if bad data were written to Redis
+    - each field is read individually, which can be inefficient and inconsistent (vs
+      getall or setall semantics)
+    """
+
+    # 2 weeks in seconds
+    DEFAULT_TTL: t.ClassVar[int] = 1209600
+
+    # required fields
+    # TODO: when `required=True` is supported in `RedisField`, set it for all of these
+    status = t.cast(TaskState, RedisField(serde=FuncxRedisEnumSerde(TaskState)))
+    internal_status = t.cast(
+        InternalTaskState, RedisField(serde=FuncxRedisEnumSerde(InternalTaskState))
+    )
+    user_id = t.cast(int, RedisField(serde=INT_SERDE))
+    function_id = t.cast(str, RedisField())
+    container = t.cast(str, RedisField())
+    task_group_id = t.cast(str, RedisField())
+    # end required fields
+
+    endpoint = t.cast(t.Optional[str], RedisField())
+
+    # FIXME: `payload` is a string which is currently being round-tripped through the
+    # JSON_SERDE. However, we cannot remove the use of the serde until we are prepared
+    # to handle the potential resulting errors. (Namely, that loading an old payload
+    # would break)
+    #
+    # consider:
+    #   >>> json.dumps("a")
+    #   '"a"'
+    #
+    # therefore, we need to at least try `json.loads` on the payload (or do a breaking
+    # change)
+    #
+    # alternatively, once `payload_reference` is populated on all tasks, we can use it
+    # to include a bool flag for how the field should be deserialized. This would
+    # require that the serde object itself have access to the payload_reference
+    payload = t.cast(t.Optional[str], RedisField(serde=JSON_SERDE))
```

Comment:
If we are planning to address the backward compatibility problem with payload using JSON_SERDE, don't we need those changes in this PR?

Reply:
No, since `payload_reference` can have a field added in the future, there's no need to solve that problem today.

In the future, we can add a field to `payload_reference`, `payload_reference["legacy_serialization"] = True`. It always gets set. We do a release. Then, we update the de/serializer for `payload` to respect `legacy_serialization` when deserializing, but still use `json.dumps` when serializing. We do a release. Then we remove `legacy_serialization` (or set it to `False`; equivalent), and stop serializing using `json.dumps`.
We can do it in fewer steps if we're willing to accept a small window during which things can break.

But the point is, it doesn't need to be solved today and it doesn't need to block this work.

I think that's important, since the serializer will need to inspect the object on which the field is attached. That's a nontrivial new feature which we shouldn't rush.

Reply:
Fair enough.